File: ihchat/app/servicos/contatos.py
"""Identidade unificada do contato entre canais."""
from __future__ import annotations

from sqlalchemy import delete, func, select, update
from sqlalchemy.orm import Session

from ..models import Contato, ContatoIdentidade, Conversa, SessaoWidget, TipoCanal
from ..util import normalizar_telefone

# canais cujo identificador ja e um dado de contato conhecido - permitem
# reconhecer que o "novo" contato e alguem que ja fala por outro canal
CANAIS_TELEFONE = {TipoCanal.WHATSAPP.value}
CANAIS_EMAIL = {TipoCanal.EMAIL.value}


def _por_identidade(sessao: Session, canal_tipo: str, identificador: str) -> Contato | None:
    return sessao.scalar(
        select(Contato)
        .join(ContatoIdentidade)
        .where(
            ContatoIdentidade.canal_tipo == canal_tipo,
            ContatoIdentidade.identificador == identificador,
        )
    )


def _por_dado_conhecido(sessao: Session, canal_tipo: str, identificador: str) -> Contato | None:
    """Reconhece o contato que ja existe com o mesmo telefone ou e-mail."""
    if canal_tipo in CANAIS_TELEFONE:
        telefone = normalizar_telefone(identificador)
        # o telefone e sempre gravado so com digitos (ver api/contatos.py),
        # entao a comparacao direta funciona em qualquer banco
        return sessao.scalar(select(Contato).where(Contato.telefone == telefone)) if telefone else None
    if canal_tipo in CANAIS_EMAIL:
        return sessao.scalar(select(Contato).where(func.lower(Contato.email) == identificador.lower()))
    return None


def normalizar_identificador(canal_tipo: str, identificador: str) -> str:
    """Como a identidade fica gravada: telefone só com dígitos, e-mail minúsculo."""
    identificador = (identificador or "").strip()
    if canal_tipo in CANAIS_TELEFONE:
        return normalizar_telefone(identificador)
    if canal_tipo in CANAIS_EMAIL:
        return identificador.lower()
    return identificador


def resolver_contato(
    sessao: Session, canal_tipo: str, identificador: str, nome_exibicao: str | None = None
) -> Contato:
    """Devolve o contato dono daquela identidade, criando-o se for a primeira vez."""
    identificador = normalizar_identificador(canal_tipo, identificador)

    contato = _por_identidade(sessao, canal_tipo, identificador)
    if contato is not None:
        if nome_exibicao and contato.nome == identificador:
            contato.nome = nome_exibicao  # antes so tinhamos o numero/endereco
        return contato

    contato = _por_dado_conhecido(sessao, canal_tipo, identificador)
    if contato is None:
        contato = Contato(
            nome=nome_exibicao or identificador,
            telefone=identificador if canal_tipo in CANAIS_TELEFONE else None,
            email=identificador if canal_tipo in CANAIS_EMAIL else None,
        )
        sessao.add(contato)
        sessao.flush()

    sessao.add(
        ContatoIdentidade(
            contato_id=contato.id,
            canal_tipo=canal_tipo,
            identificador=identificador,
            nome_exibicao=nome_exibicao,
        )
    )
    sessao.flush()
    sessao.refresh(contato)
    return contato


def identificador_no_canal(sessao: Session, contato: Contato, canal_tipo: str) -> str | None:
    """Uma identidade do contato no canal (a mais antiga).

    Para responder a uma conversa use mensagens.destino_da_conversa, que
    prefere o número que escreveu NELA: depois de uma mesclagem o contato
    pode ter dois números no mesmo canal, e "o primeiro" seria o de outra
    conversa — talvez de outra pessoa.
    """
    identidade = sessao.scalar(
        select(ContatoIdentidade)
        .where(
            ContatoIdentidade.contato_id == contato.id,
            ContatoIdentidade.canal_tipo == canal_tipo,
        )
        .order_by(ContatoIdentidade.id)
        .limit(1)
    )
    if identidade:
        return identidade.identificador
    if canal_tipo in CANAIS_TELEFONE and contato.telefone:
        return normalizar_telefone(contato.telefone)
    if canal_tipo in CANAIS_EMAIL and contato.email:
        return contato.email
    return None


def mesclar(sessao: Session, principal: Contato, secundario: Contato) -> Contato:
    """Junta um contato duplicado no principal, preservando todo o historico.

    As sessões do widget dos DOIS contatos são ENCERRADAS, não transferidas:
    mesclar é uma decisão da atendente, não uma prova de identidade, e o
    histórico da ficha acabou de ganhar as conversas (e as respostas) da
    outra. O widget recebe 401, abre uma sessão nova e o visitante continua
    conversando; a equipe segue vendo tudo unificado no painel.
    """
    if principal.id == secundario.id:
        return principal

    # comandos diretos, e nao a colecao do ORM: mexer na colecao do contato que
    # sera apagado faria o cascade delete-orphan levar junto as identidades
    existentes = {(i.canal_tipo, i.identificador) for i in principal.identidades}
    duplicadas = [
        i.id for i in secundario.identidades if (i.canal_tipo, i.identificador) in existentes
    ]
    if duplicadas:
        sessao.execute(delete(ContatoIdentidade).where(ContatoIdentidade.id.in_(duplicadas)))
    sessao.execute(
        update(ContatoIdentidade)
        .where(ContatoIdentidade.contato_id == secundario.id)
        .values(contato_id=principal.id)
    )
    sessao.execute(
        update(Conversa).where(Conversa.contato_id == secundario.id).values(contato_id=principal.id)
    )
    # os dois lados: nenhum navegador herda o histórico da outra ficha
    sessao.execute(delete(SessaoWidget).where(SessaoWidget.contato_id.in_([principal.id, secundario.id])))
    principal.email = principal.email or secundario.email
    principal.telefone = principal.telefone or secundario.telefone
    principal.empresa = principal.empresa or secundario.empresa
    principal.documento = principal.documento or secundario.documento
    # observações das duas fichas ficam juntas (o e-mail que o visitante
    # digitou no widget, por exemplo, está nelas)
    obs_p, obs_s = (principal.observacoes or "").strip(), (secundario.observacoes or "").strip()
    if obs_s and obs_s != obs_p:
        principal.observacoes = f"{obs_p}\n\n{obs_s}" if obs_p else obs_s
    sessao.flush()
    sessao.expire(secundario)
    sessao.delete(secundario)
    sessao.flush()
    sessao.refresh(principal)
    return principal

File: omnichannel/php/app/Atendimento/Contatos.php
<?php
declare(strict_types=1);

namespace OmniChannel\Atendimento;

use OmniChannel\Banco\Banco;
use OmniChannel\Nucleo\Datas;
use OmniChannel\Nucleo\Texto;

/**
 * Identidade unificada do contato entre canais (app/servicos/contatos.py).
 *
 * Um contato pode falar por vários canais: a identidade externa (número de
 * WhatsApp, chat_id do Telegram, e-mail) fica em contato_identidades, e não
 * no contato. Assim o histórico do cliente continua único quando ele troca
 * de canal.
 */
final class Contatos
{
    /** Canais cujo identificador já é um dado de contato conhecido. */
    public const CANAIS_TELEFONE = ['whatsapp'];
    public const CANAIS_EMAIL = ['email'];

    /** @return array<string, mixed>|null a linha do contato */
    public static function porId(int $id): ?array
    {
        return Banco::um('SELECT * FROM contatos WHERE id = ?', [$id]);
    }

    /** Como a identidade é gravada: WhatsApp só com dígitos, e-mail em minúsculas. */
    public static function normalizarIdentificador(string $canalTipo, string $identificador): string
    {
        $identificador = trim($identificador);
        if (in_array($canalTipo, self::CANAIS_TELEFONE, true)) {
            return Texto::normalizarTelefone($identificador);
        }
        if (in_array($canalTipo, self::CANAIS_EMAIL, true)) {
            return mb_strtolower($identificador);
        }
        return $identificador;
    }

    /**
     * O contato dono daquela identidade, criando-o na primeira vez.
     *
     * Ordem: identidade já conhecida → contato com o mesmo telefone (WhatsApp)
     * ou e-mail (e-mail) → contato novo. Devolve o id.
     */
    public static function resolver(string $canalTipo, string $identificador, ?string $nomeExibicao = null): int
    {
        return Concorrencia::transacao(static function () use ($canalTipo, $identificador, $nomeExibicao): int {
            $identificador = self::normalizarIdentificador($canalTipo, $identificador);
            $nomeExibicao = $nomeExibicao === null || trim($nomeExibicao) === '' ? null : mb_substr(trim($nomeExibicao), 0, 160);

            $existente = self::porIdentidade($canalTipo, $identificador);
            if ($existente !== null) {
                if ($nomeExibicao !== null && (string) $existente['nome'] === $identificador) {
                    // antes só tínhamos o número/endereço
                    Banco::atualizar('contatos', ['nome' => $nomeExibicao, 'atualizado_em' => Datas::agoraBanco()], 'id = ?', [$existente['id']]);
                }
                return (int) $existente['id'];
            }

            $contatoId = self::porDadoConhecido($canalTipo, $identificador);
            $criado = false;
            if ($contatoId === null) {
                $agora = Datas::agoraBanco();
                $contatoId = Banco::inserir('contatos', [
                    'nome' => mb_substr($nomeExibicao ?? $identificador, 0, 160),
                    'telefone' => in_array($canalTipo, self::CANAIS_TELEFONE, true) ? mb_substr($identificador, 0, 32) : null,
                    'email' => in_array($canalTipo, self::CANAIS_EMAIL, true) ? mb_substr($identificador, 0, 160) : null,
                    'criado_em' => $agora,
                    'atualizado_em' => $agora,
                ]);
                $criado = true;
            }
            try {
                Banco::inserir('contato_identidades', [
                    'contato_id' => $contatoId,
                    'canal_tipo' => $canalTipo,
                    'identificador' => mb_substr($identificador, 0, 200),
                    'nome_exibicao' => $nomeExibicao,
                    'criado_em' => Datas::agoraBanco(),
                ]);
            } catch (\PDOException $erro) {
                // duas entregas simultâneas do mesmo contato novo: a outra
                // gravou primeiro; fica valendo a dela. A releitura é TRAVADA:
                // no MySQL uma leitura comum veria a foto do início desta
                // transação, sem a identidade que a outra acabou de confirmar
                if (!Banco::eUnicidade($erro)) {
                    throw $erro;
                }
                $vencedor = self::porIdentidade($canalTipo, $identificador, travar: true);
                if ($vencedor === null) {
                    throw $erro;
                }
                if ($criado) {
                    Banco::executar('DELETE FROM contatos WHERE id = ?', [$contatoId]);
                }
                return (int) $vencedor['id'];
            }
            return $contatoId;
        });
    }

    /**
     * Para onde responder este contato dentro de um canal (null: não há como).
     *
     * É o ÚLTIMO recurso: a resposta de uma conversa vai para a identidade
     * que escreveu nela (Mensagens::destinoDaConversa). Depois de uma
     * mesclagem o contato pode ter dois números no mesmo canal, e "o
     * primeiro" seria o de outra conversa — talvez de outra pessoa.
     */
    public static function identificadorNoCanal(int $contatoId, string $canalTipo): ?string
    {
        $identificador = Banco::valor(
            'SELECT identificador FROM contato_identidades WHERE contato_id = ? AND canal_tipo = ? ORDER BY id LIMIT 1',
            [$contatoId, $canalTipo]
        );
        if ($identificador !== null) {
            return (string) $identificador;
        }
        $contato = self::porId($contatoId);
        if ($contato === null) {
            return null;
        }
        if (in_array($canalTipo, self::CANAIS_TELEFONE, true) && (string) ($contato['telefone'] ?? '') !== '') {
            return Texto::normalizarTelefone((string) $contato['telefone']);
        }
        if (in_array($canalTipo, self::CANAIS_EMAIL, true) && (string) ($contato['email'] ?? '') !== '') {
            return (string) $contato['email'];
        }
        return null;
    }

    /**
     * Junta um contato duplicado no principal, sem perder nada: identidades e
     * conversas passam para o principal; dados que só o secundário tinha
     * completam a ficha. Devolve o id do principal.
     *
     * As sessões do widget dos DOIS contatos são ENCERRADAS, não
     * transferidas: mesclar é uma decisão da atendente, não uma prova de
     * identidade. O widget mostra ao navegador o histórico do contato ligado
     * à sessão, e esse histórico acabou de ganhar as conversas (e as
     * identidades, ou seja, as mensagens futuras) da outra ficha. Encerrar só
     * a do secundário (como o ON DELETE CASCADE do Python) deixava o caminho
     * inverso aberto: o visitante anônimo que diz "sou a Maria do e-mail",
     * mesclado como PRINCIPAL, passava a ler o e-mail e o WhatsApp da Maria,
     * e dois visitantes do mesmo site juntados liam um a conversa do outro.
     * O widget recebe 401, abre uma sessão nova e o visitante continua
     * conversando; a equipe segue vendo tudo unificado no painel.
     */
    public static function mesclar(int $principalId, int $secundarioId): int
    {
        if ($principalId === $secundarioId) {
            return $principalId;
        }
        return Concorrencia::transacao(static function () use ($principalId, $secundarioId): int {
            // trava os dois contatos (sempre na mesma ordem) antes de mexer:
            // uma mensagem chegando para um deles espera a mesclagem terminar
            foreach ([min($principalId, $secundarioId), max($principalId, $secundarioId)] as $id) {
                Banco::valor('SELECT id FROM contatos WHERE id = ?' . Concorrencia::travando(), [$id]);
            }
            $principal = self::porId($principalId);
            $secundario = self::porId($secundarioId);
            if ($principal === null || $secundario === null) {
                throw new \RuntimeException('contato nao encontrado');
            }
            // identidade que os dois têm fica só a do principal (a unicidade
            // canal+identificador impediria a mudança de dono)
            $existentes = [];
            foreach (Banco::todos('SELECT canal_tipo, identificador FROM contato_identidades WHERE contato_id = ?', [$principalId]) as $i) {
                $existentes[$i['canal_tipo'] . "\0" . $i['identificador']] = true;
            }
            foreach (Banco::todos('SELECT id, canal_tipo, identificador FROM contato_identidades WHERE contato_id = ?', [$secundarioId]) as $i) {
                if (isset($existentes[$i['canal_tipo'] . "\0" . $i['identificador']])) {
                    Banco::executar('DELETE FROM contato_identidades WHERE id = ?', [(int) $i['id']]);
                }
            }
            Banco::executar('UPDATE contato_identidades SET contato_id = ? WHERE contato_id = ?', [$principalId, $secundarioId]);
            $movidas = array_map('intval', array_column(
                Banco::todos('SELECT id FROM conversas WHERE contato_id = ?', [$secundarioId]),
                'id'
            ));
            Banco::executar('UPDATE conversas SET contato_id = ? WHERE contato_id = ?', [$principalId, $secundarioId]);
            // os dois lados (ver o comentário da função): nenhum navegador
            // herda o histórico da outra ficha
            Banco::executar('DELETE FROM sessoes_widget WHERE contato_id IN (?, ?)', [$principalId, $secundarioId]);

            $mudancas = [];
            foreach (['email', 'telefone', 'empresa', 'documento'] as $campo) {
                if ((string) ($principal[$campo] ?? '') === '' && (string) ($secundario[$campo] ?? '') !== '') {
                    $mudancas[$campo] = $secundario[$campo];
                }
            }
            $obsP = trim((string) ($principal['observacoes'] ?? ''));
            $obsS = trim((string) ($secundario['observacoes'] ?? ''));
            if ($obsS !== '' && $obsS !== $obsP) {
                $mudancas['observacoes'] = $obsP === '' ? $obsS : $obsP . "\n\n" . $obsS;
            }
            $mudancas['atualizado_em'] = Datas::agoraBanco();
            Banco::atualizar('contatos', $mudancas, 'id = ?', [$principalId]);
            Banco::executar('DELETE FROM contatos WHERE id = ?', [$secundarioId]);

            // o painel atualiza a ficha nas conversas que mudaram de dono
            foreach ($movidas as $conversaId) {
                Conversas::publicar($conversaId);
            }
            return $principalId;
        });
    }

    /** @return array<string, mixed>|null */
    private static function porIdentidade(string $canalTipo, string $identificador, bool $travar = false): ?array
    {
        return Banco::um(
            'SELECT c.* FROM contatos c JOIN contato_identidades i ON i.contato_id = c.id
             WHERE i.canal_tipo = ? AND i.identificador = ? ORDER BY i.id LIMIT 1'
            . ($travar ? Concorrencia::travando() : ''),
            [$canalTipo, $identificador]
        );
    }

    /** O contato que já existe com o mesmo telefone ou e-mail. */
    private static function porDadoConhecido(string $canalTipo, string $identificador): ?int
    {
        if (in_array($canalTipo, self::CANAIS_TELEFONE, true)) {
            $telefone = Texto::normalizarTelefone($identificador);
            if ($telefone === '') {
                return null;
            }
            // o telefone é sempre gravado só com dígitos (ver PATCH do contato)
            $id = Banco::valor('SELECT id FROM contatos WHERE telefone = ? ORDER BY id LIMIT 1', [$telefone]);
            return $id === null ? null : (int) $id;
        }
        if (in_array($canalTipo, self::CANAIS_EMAIL, true)) {
            $id = Banco::valor('SELECT id FROM contatos WHERE LOWER(email) = ? ORDER BY id LIMIT 1', [mb_strtolower($identificador)]);
            return $id === null ? null : (int) $id;
        }
        return null;
    }
}
